docs/codeql/ql-language-reference/types.rst

```diff
@@ -382,6 +384,68 @@ from ``OneTwoThree`` and ``int``.
    must :ref:`override <overriding-member-predicates>` those definitions to avoid ambiguity.
    :ref:`Super expressions <super>` are often useful in this situation.
 
+.. _final-extensions:
+
+Final extensions
+================
+
+A class can extend final types or final aliases of types. In that case, it inherits final
+versions of all the member predicates and fields of those supertypes.
+Member predicates that are inherited through final extensions cannot be overridden,
+but they can be shadowed.
+
+For example, extending the class from the :ref:`first example <defining-a-class>`:
+
+.. code-block:: ql
+
+    final class FinalOneTwoThree = OneTwoThree;
+    
+    class OneTwo extends FinalOneTwoThree {
+      OneTwo() {
+        this = 1 or this = 2
+      }
+    
+      string getAString() {
+        result = "One or two: " + this.toString()
+      }
+    }
+
+The member predicate ``getAString()`` shadows the original definition of ``getAString()``
+from ``OneTwoThree``.
+
+Different to overriding (see ":ref:`overriding-member-predicates`"),
+final extensions leave the extended type unchanged:
+
+.. code-block:: ql
+
+    from OneTwoTree o
+    select o, o.getAString()
+
++---+-------------------------+
+| o | ``getAString()`` result |
++===+=========================+
+| 1 | One, two or three: 1    |
++---+-------------------------+
+| 2 | One, two or three: 2    |
++---+-------------------------+
+| 3 | One, two or three: 3    |
++---+-------------------------+
+
+However, when calling ``getAString()`` on ``OneTwo``, the original definition is shadowed:
+
+.. code-block:: ql
+
+    from OneTwo o
+    select o, o.getAString()
+
++---+-------------------------+
+| o | ``getAString()`` result |
++===+=========================+
+| 1 | One or two: 1             |
++---+-------------------------+
+| 2 | One or two: 2             |
```

Comment:
```suggestion
| 1 | One or two: 1           |
+---+-------------------------+
| 2 | One or two: 2           |
```

Reply:
I have force-pushed a commit that fixes this.